### DISCUSSION TITLE
mirthDB extracting meta data about channel

### DISCUSSION THREAD

**Discussion by aungktun** [nextgenhealthcare/connect#5245] [Q&A]
We have mirth with MS SQL and there are tables in mirthdb (Channel, Channel_Group and Configuration).

Is there a sql already written to join all of the 3 tables and get the data out of xml fields . For example channel_group has field called channel_group which contains xml data about which channel are in this channel_group. 

Also in CONFIGURATION table there is a field configuration.name with value 'channelMetadata' and configuration.value has xml data which contains additional meta information about a particular channel which is usually display in Summary tab.

How do I get/extract all the information using sql so we can create table about channel/job and its meta data without hardcoding? 

Here is sample sql I started. My sql code for xml extraction is base on https://stackoverflow.com/questions/19165213/how-to-query-for-xml-values-and-attributes-from-table-in-sql-server

`SELECT ID, NAME, REVISION, CHANNEL_GROUP, convert(xml, CHANNEL_GROUP) as CHANNEL_GROUP_XML
	into #channel_group
	FROM mirthDB.dbo.CHANNEL_GROUP

	select id, name, revision, channel 
	, convert(xml, channel) as channel_xml
	into #channel
	from mirthDB.dbo.CHANNEL as c 

	--drop table #channel_data_extracted
	select c.*
	, ch.i.value('./description[1]', 'varchar(8000)') as channel_Description
	--, cnt.m.value('./pollConnectorProperties[1]', 'varchar(4000)') as pollConnectorProperties
	, cnt.m.value('./pollConnectorProperties[1]/pollingType[1]', 'varchar(40)') as pollingType
	, cnt.m.value('./pollConnectorProperties[1]/pollOnStart[1]', 'varchar(40)') as pollOnStart
	, cnt.m.value('./pollConnectorProperties[1]/pollingFrequency[1]', 'varchar(40)') as pollingFrequency
	, cnt.m.value('./pollConnectorProperties[1]/pollingHour[1]', 'varchar(40)') as pollingHour
	, cnt.m.value('./pollConnectorProperties[1]/pollingMinute[1]', 'varchar(40)') as pollingMinute
	, cnt.m.value('./pollConnectorProperties[1]/cronJobs[1]', 'varchar(40)') as cronJobs
	, CASE WHEN cnt.m.value('./pollConnectorProperties[1]/cronJobs[1]', 'varchar(40)') != '' THEN 
			cnt.m.value('./pollConnectorProperties[1]/cronJobs[1]/cronProperty[1]/expression[1]', 'varchar(400)') else '' end as cronJobs_expression_1
	, CASE WHEN cnt.m.value('./pollConnectorProperties[1]/cronJobs[1]', 'varchar(40)') != '' THEN 
			cnt.m.value('./pollConnectorProperties[1]/cronJobs[1]/cronProperty[2]/expression[1]', 'varchar(400)') else '' end as cronJobs_expression_2          
	--, cnt.m.value('./pollConnectorProperties[1]/pollConnectorPropertiesAdvanced[1]', 'varchar(4000)') as pollConnectorPropertiesAdvanced
	, cnt.m.value('./pollConnectorProperties[1]/pollConnectorPropertiesAdvanced[1]/weekly[1]', 'varchar(40)') as weekly
	--, cnt.m.value('./pollConnectorProperties[1]/pollConnectorPropertiesAdvanced[1]/inactiveDays[1]', 'varchar(40)') as inactiveDays
	, cnt.m.value('./pollConnectorProperties[1]/pollConnectorPropertiesAdvanced[1]/inactiveDays[1]/boolean[1]', 'varchar(40)') as inactiveDays_Mon
	, cnt.m.value('./pollConnectorProperties[1]/pollConnectorPropertiesAdvanced[1]/inactiveDays[1]/boolean[2]', 'varchar(40)') as inactiveDays_Tue
	, cnt.m.value('./pollConnectorProperties[1]/pollConnectorPropertiesAdvanced[1]/inactiveDays[1]/boolean[3]', 'varchar(40)') as inactiveDays_Wed
	, cnt.m.value('./pollConnectorProperties[1]/pollConnectorPropertiesAdvanced[1]/inactiveDays[1]/boolean[4]', 'varchar(40)') as inactiveDays_Thu
	, cnt.m.value('./pollConnectorProperties[1]/pollConnectorPropertiesAdvanced[1]/inactiveDays[1]/boolean[5]', 'varchar(40)') as inactiveDays_Fri
	, cnt.m.value('./pollConnectorProperties[1]/pollConnectorPropertiesAdvanced[1]/inactiveDays[1]/boolean[6]', 'varchar(40)') as inactiveDays_Sat
	, cnt.m.value('./pollConnectorProperties[1]/pollConnectorPropertiesAdvanced[1]/inactiveDays[1]/boolean[7]', 'varchar(40)') as inactiveDays_Sun
	, cnt.m.value('./pollConnectorProperties[1]/pollConnectorPropertiesAdvanced[1]/dayOfMonth[1]', 'varchar(40)') as dayOfMonth
	, cnt.m.value('./pollConnectorProperties[1]/pollConnectorPropertiesAdvanced[1]/allDay[1]', 'varchar(40)') as allDay
	, cnt.m.value('./pollConnectorProperties[1]/pollConnectorPropertiesAdvanced[1]/startingHour[1]', 'varchar(40)') as startingHour
	, cnt.m.value('./pollConnectorProperties[1]/pollConnectorPropertiesAdvanced[1]/startingMinute[1]', 'varchar(40)') as startingMinute
	, cnt.m.value('./pollConnectorProperties[1]/pollConnectorPropertiesAdvanced[1]/endingHour[1]', 'varchar(40)') as endingHour
	, cnt.m.value('./pollConnectorProperties[1]/pollConnectorPropertiesAdvanced[1]/endingMinute[1]', 'varchar(40)') as endingMinute
	into #channel_data_extracted
	from #channel as c 
	cross apply c.channel_xml.nodes( '/channel/sourceConnector/properties' ) cnt(m)
	cross apply c.channel_xml.nodes( '/channel' ) ch(i)

	--select * from #channel_data_extracted

	SELECT [CATEGORY]
		  ,[NAME]
		  ,convert(xml, [VALUE] ) as xml_value
	into #config
	FROM [mirthDB].[dbo].[CONFIGURATION]
	where name = 'channelMetadata'

	--drop table #channel_group_extracted
	select
	xt.id as channel_group_id
	, xt.name  as channel_group_name
	, xt.revision  as channel_group_revision
	, xt.channel_group 
	, xt.channel_group_xml
	, id.m.value( '.', 'varchar(50)' ) channel_id
	, time.m.value( '.', 'varchar(50)' ) time
	, timezone.m.value( '.', 'varchar(50)' ) timezone
	into #channel_group_extracted
	FROM #channel_group as xt
	cross apply xt.CHANNEL_GROUP_XML.nodes( '/channelGroup/channels/channel/id' ) id(m)
	cross apply xt.CHANNEL_GROUP_XML.nodes( '/channelGroup/lastModified/time' ) time(m)
	cross apply xt.CHANNEL_GROUP_XML.nodes( '/channelGroup/lastModified/timezone' ) timezone(m)

	--drop table #config_channel_metadata
	select distinct ent.m.value('./string[1]', 'varchar(100)') as channel_id
	, ent.m.value('./com.mirth.connect.model.ChannelMetadata[1]/enabled[1]', 'varchar(10)') as channel_enable_ind
	into #config_channel_metadata
	from #config as cfg
	cross apply cfg.xml_value.nodes( 'map/entry' ) ent(m)

	select dc.LOCAL_CHANNEL_ID, dc.CHANNEL_ID
	, c.name, c.revision
	, c.channel_Description
	, c.pollingType,c.pollOnStart,c.pollingFrequency,c.pollingHour
	, c.pollingMinute,c.cronJobs,c.cronJobs_expression_1,c.cronJobs_expression_2,c.weekly
	, c.inactiveDays_Mon,c.inactiveDays_Tue,c.inactiveDays_Wed,c.inactiveDays_Thu,c.inactiveDays_Fri
	, c.inactiveDays_Sat,c.inactiveDays_Sun,c.dayOfMonth,c.allDay,c.startingHour,c.startingMinute
	, c.endingHour,c.endingMinute
	, cg.channel_group_name
	, ccm.channel_enable_ind
	from mirthdb.dbo.D_CHANNELS dc 
	left join #channel_data_extracted c on c.ID = dc.CHANNEL_ID
	left join #channel_group_extracted as cg on cg.channel_id = c.id
	left join #config_channel_metadata as ccm on ccm.channel_id = dc.CHANNEL_ID
	order by cg.channel_group_name,ccm.channel_enable_ind desc, c.NAME`

**Comment by jonbartels** [comment]
Some references that might help are:
* https://gist.github.com/MichaelLeeHobbs/67980d165fc68880eb2ab283c673244b - this extracts table size info about a channel from postgres it may help show what to query, you'll have to convert it for SQL Server though
* https://github.com/kayyagari/connect/blob/je/mc-integ-tests/mc-db-tables.png - this is a visualization of the MC DB tables and can help find data

The challenge that I think you're running into is that each channel connector type will have its own properties. You did a good job digging into the poller connector for all those details. You'll have to do that again for the HTTP connector, then again for TCP, and so on. There are only about 10 connector types but you'll have to grind through each one. A better approach might be to do something like read the connector properties, then produce output where you just return the properties XML as a blob or iterate over each child and return a SQL array or a comma-separated list of properties for each connector type.

If you are who I think you are, you should check with both Zen and NextGen. You work for a customer of both vendors and each of them offers a channel reporting tool that operates off of a server export from MC.

**Reply by aungktun** [reply]
Hi Jon,
Thanks. Yes. I am who you are thinking from SMCH. We have other MC instances for other interfaces outside of HIE.
